llvm/lib/SYCLLowerIR/ComputeModuleRuntimeInfo.cpp

```diff
@@ -161,14 +164,59 @@ std::optional<T> getKernelSingleEltMetadata(const Function &Func,
 
 PropSetRegTy computeModuleProperties(const Module &M,
                                      const EntryPointSet &EntryPoints,
-                                     const GlobalBinImageProps &GlobProps) {
+                                     const GlobalBinImageProps &GlobProps,
+                                     const StringRef &DeviceLibSPVLoc) {
 
   PropSetRegTy PropSet;
-  {
-    uint32_t MRMask = getSYCLDeviceLibReqMask(M);
-    std::map<StringRef, uint32_t> RMEntry = {{"DeviceLibReqMask", MRMask}};
-    PropSet.add(PropSetRegTy::SYCL_DEVICELIB_REQ_MASK, RMEntry);
+  // If sycl-post-link doesn't specify a valid fallback spv path, the param
+  // 'DeviceLibSPVLoc' is set to an empty string.
+  if (!DeviceLibSPVLoc.empty()) {
+    SmallVector<llvm::SYCLDeviceLibSPVMeta, 16> RequiredLibs;
+    llvm::getRequiredSYCLDeviceLibs(M, RequiredLibs);
+    for (auto ExtMeta : RequiredLibs) {
+      std::string SPVPath =
+          DeviceLibSPVLoc.str() + "/" + std::string(ExtMeta.SPVFileName);
+      if (!llvm::sys::fs::exists(SPVPath))
+        continue;
+
+      auto SPVMB = llvm::MemoryBuffer::getFile(SPVPath);
```

Comment:
I have a couple of concerns here:
1. Final binary size. If we do per-module embedding of devicelib binaries, it means that we may end up embedding devicelib multiple times, i.e. once per device image that uses it. I'm afraid that it may have drastic impact on final binary size when per-kernel device code split is involved.
2. Link-time performance. We seem to read devicelib file for every split module we produce over and over again.

Ideally, we should re-use shared libraries support infrastructure for this, i.e.:
- we pass `devicelib` object files to link step as if they were regular user-provided object files
- we outline `devicelib` into separate device images using our existing rules for shared libraries. As part of that effort, every symbol provided by a library will be listed as exported
  - [new] we may skip `__devicelib` symbols when computing list of exports right now, that will need to be changed if that's the case
- [new] we consider `__devicelib` symbols when computing list of imports for a device image
- [new] at SYCL RT, we *move* a mechanism of devicelib linking and integrate it with shared libraries support:
  - every `__devicelib` import is handled differently from regular imports. Instead of just unconditionally linking a device image which provides that symbol, we first check if backend can supply us with it and if so, we don't use our embedded fallback devicelib

Shared libraries design doc can be found [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/SharedLibraries.md)

Reply:
Hi, @AlexeySachkov 
"If we do per-module embedding of devicelib binaries, it means that we may end up embedding devicelib multiple times"
We don't do per-module embedding, the embedding happens only in sycl-post-link phase, all input modules have been linked together to a full LLVM IR and we analyze the full LLVM IR file to figure out which spv files are required. For example, if an application has module A, B, C, all use assert function, after llvm-link links them all to a full module ABC.bc and pass it to sycl-post-link tool, sycl-post-link will figure out ABC.bc depend on libsycl-fallback-cassert.spv and embed it to final device image for only 1 time.

For Link time performance, when SYCL runtime checks the embedded spv and extract when required, we will record which spv modules have been extracted, no multiple read/extract will happen.

For most cases which don't use devicelib functions, there will be no code-size/link-time performance since sycl-post-link analysis shows no dependency on any device function and sycl-post-link and sycl runtime will just skip the embedding and extracting work.

Thanks very much.

Reply:
> For Link time performance, when SYCL runtime checks the embedded spv and extract when required, we will record which spv modules have been extracted, no multiple read/extract will happen.

Here I meant link operation which happens during the app build, i.e. if we have multiple device images which use the same devicelib, then we will read that `.spv` file multiple times in `sycl-post-link` (once per device image) if I understand correctly.

> Do you mean we should move all required spv into a new device image together with user's device image and link it with any user's image which depends on it?

Yes, exactly. We already have an infrastructure to record imported/exported symbols information into device images. SYCL RT already has a functionality to "resolve" symbols by linking several device images together based on that imported/exported symbols list. However, `devicelib` had been implemented way before we got support for that and it uses separate mechanism.

Right now we specifically ignore everything which starts with `__` when recording imported/exported symbols, though.

The idea which I would like to propose and discuss is that we do consider `__devicelib_*` symbols as imported/exported and update our SYCL RT logic to be more careful with those: we only need to link a devicelib fallback image if corresponding extension is *not* supported by a device. I assume that it will also require to move function -> library mapping that we have in the compiler into the runtime, so that runtime knows which extension it should check for to see if a fallback library should be linked to resolve a `__devicelib_*` function.

Reply:
Hi, @AlexeySachkov 
I think your suggestion to make required fallback spv to be "device images" and bundle together with user's device image does make sense. I will try this approach for the first step, this should help us to avoid reading/embedding spv data for multiple times.
Thanks very much.